libraries/rush-lib/src/logic/operations/IPCOperationRunner.ts

```diff
@@ -0,0 +1,210 @@
+              resolveReadyPromise();
+            }
+          });
+        } else {
+          terminal.writeLine(`Connecting to existing IPC process...`);
+        }
+        const subProcess: ChildProcess = this._ipcProcess;
+        let hasWarningOrError: boolean = false;
+
+        function onStdout(data: Buffer): void {
+          const text: string = data.toString();
+          terminalProvider.write(text, TerminalProviderSeverity.log);
+        }
+        function onStderr(data: Buffer): void {
+          const text: string = data.toString();
+          terminalProvider.write(text, TerminalProviderSeverity.error);
+          hasWarningOrError = true;
+        }
+
+        // Hook into events, in order to get live streaming of the log
+        subProcess.stdout?.on('data', onStdout);
+        subProcess.stderr?.on('data', onStderr);
+
+        const status: OperationStatus = await new Promise((resolve, reject) => {
+          function finishHandler(message: unknown): void {
+            if (isAfterExecuteEventMessage(message)) {
+              terminal.writeLine('Received finish notification');
+              subProcess.stdout?.off('data', onStdout);
+              subProcess.stderr?.off('data', onStderr);
+              subProcess.off('message', finishHandler);
+              subProcess.off('error', reject);
+              subProcess.off('exit', onExit);
+              terminal.writeLine('Disconnected from IPC process');
+              // These types are currently distinct but have the same underlying values
+              resolve(message.status as unknown as OperationStatus);
+            }
+          }
+
+          function onExit(code: number): void {
+            try {
+              if (code !== 0) {
+                // Do NOT reject here immediately, give a chance for other logic to suppress the error
+                context.error = new OperationError('error', `Returned error code: ${code}`);
+                resolve(OperationStatus.Failure);
+              } else if (hasWarningOrError) {
+                resolve(OperationStatus.SuccessWithWarning);
+              } else {
+                resolve(OperationStatus.Success);
+              }
+            } catch (error) {
+              reject(error as OperationError);
+            }
+          }
+
+          subProcess.on('message', finishHandler);
+          subProcess.on('error', reject);
+          subProcess.on('exit', onExit);
+
+          this._processReadyPromise!.then(() => {
+            isConnected = true;
+            terminal.writeLine('Child supports IPC protocol. Sending "run" command...');
+            const runCommand: IRunCommandMessage = {
+              command: 'run'
+            };
+            subProcess.send(runCommand);
+          }, reject);
+        });
+
+        if (isConnected && !this._persist) {
+          await this.shutdownAsync();
+        }
+
+        // @rushstack/operation-graph does not currently have a concept of "Success with Warning"
```

Comment:
How difficult would it be to add that?

Reply:
We'd have to add it to the enumeration, then figure out what it means in the context of Heft, etc. Probably not terribly difficult, but the logic here matches existing `ShellOperationRunner` semantics.

Easy enough to do as a follow-on PR, since `@rushstack/operation-graph` will need it anyway before it can replace Rush's engine.